src/browser/modules/Editor/Monaco.tsx

```diff
@@ -41,193 +51,480 @@ import {
-        ?.getLinesContent()
-        .join('\n') || ''
-
-    onChange(text)
-    addWarnings(parse(text).referencesListener.queriesAndCommands)
-  }
+    // The monaco render method does not redraw line numbers
+    // Getting and setting content and cursor is done to force a redraw
+    useEffect(() => {
+      const cursorPosition = editorRef?.current?.getPosition() as IPosition
+      editorRef.current?.setValue(editorRef.current?.getValue() || '')
+      editorRef.current?.setPosition(cursorPosition)
+    }, [useDb])
 
-  const debouncedUpdateCode = debounce(updateCode, 300)
+    // Create monaco instance, listen to text changes and destroy
+    useEffect(() => {
+      languages.register({ id: 'cypher' })
+      languages.setTokensProvider('cypher', new CypherTokensProvider())
 
-  // On each text change, clear warnings and reset countdown to adding warnings
-  const onContentUpdate = () => {
-    editor.setModelMarkers(
-      editorRef.current?.getModel() as editor.ITextModel,
-      monacoId,
-      []
-    )
-    debouncedUpdateCode()
-  }
+      editorSupportRef.current = new CypherEditorSupport(value)
+      if (schema) {
+        editorSupportRef.current.setSchema(schema)
+      }
 
-  const addWarnings = (
-    statements: { start: { line: number }; getText: () => string }[]
-  ) => {
-    if (!statements.length) return
+      languages.registerCompletionItemProvider('cypher', {
+        triggerCharacters: ['.', ':', '[', '(', '{'],
+        provideCompletionItems: (model, position) => {
+          var { startColumn, endColumn } = model.getWordUntilPosition(position)
+          const range = {
+            startLineNumber: position.lineNumber,
+            endLineNumber: position.lineNumber,
+            startColumn,
+            endColumn
+          }
+          editorSupportRef.current?.update(editorRef.current?.getValue() || '')
+          const items =
+            editorSupportRef.current?.getCompletion(
+              position.lineNumber,
+              position.column
+            ).items || []
 
-    const model = editorRef.current?.getModel() as editor.ITextModel
+          return {
+            suggestions: items.map(item => ({
+              label: item.view,
+              kind: languages.CompletionItemKind.Keyword,
+              insertText: item.content,
+              range
+            }))
+          }
+        }
+      })
 
-    // clearing markers again solves issue with incorrect multi-statement warning when user spam clicks setting on and off
-    editor.setModelMarkers(
-      editorRef.current?.getModel() as editor.ITextModel,
-      monacoId,
-      []
-    )
+      editor.defineTheme(DARK_THEME, monacoDarkTheme)
+      editor.defineTheme(LIGHT_THEME, monacoLightTheme)
+      editor.defineTheme(OUTLINE_THEME, monacoLightTheme)
+      // Browser's light theme is called 'normal', but OS's light theme is called 'light'
+      // 'light' is used when theme is set to light in OS and auto in browser
```

Comment:
(sidenote) Perhaps we should call it light as well?

Reply:
I agree, I don't see a reason for the value to be normal